### PR TITLE
feat!: initialize `Serializer` with capacity

### PR DESCRIPTION
**Pull request by tbrezot**
Use the new method `<impl Serializable>::length()` to initialize the `Serializer` in `<impl Serializable>::try_to_bytes()` with the correct capacity.

BREAKING CHANGE: do not write the size of the `ciphertext` of the `plaintext` in the serialization of the `EncryptedHeader` and `PlaintextHeader` respectively.

**Note**: will replace the `rev` in `Cargo.toml` once [`feat/serializer_with_capacity`](https://github.com/Cosmian/crypto_core/pull/26) is merged on CryptoCore

**Note**: maybe the `Option<usize>` in the signature of `<impl Serializable>::length()` can be changed into `usize` and the implementation of this method be forced.

## Benchmarks
CPU used: Intel(R) Core(TM) i5-6200U CPU @ 2.30GHz.
```
key serialization/MSK   time:   [324.29 ns 325.40 ns 326.68 ns]
                        change: [-25.235% -24.363% -23.303%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 322 outliers among 5000 measurements (6.44%)
  124 (2.48%) high mild
  198 (3.96%) high severe
key serialization/MPK   time:   [95.324 µs 95.442 µs 95.565 µs]
                        change: [-0.1535% +0.0186% +0.1876%] (p = 0.84 > 0.05)
                        No change in performance detected.
Found 89 outliers among 5000 measurements (1.78%)
  25 (0.50%) high mild
  64 (1.28%) high severe
key serialization/USK   time:   [89.893 ns 90.211 ns 90.573 ns]
                        change: [-26.650% -26.100% -25.511%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 220 outliers among 5000 measurements (4.40%)
  75 (1.50%) high mild
  145 (2.90%) high severe

header serialization/1 partition
                        time:   [11.364 µs 11.379 µs 11.396 µs]
                        change: [-0.0646% +0.1125% +0.2933%] (p = 0.24 > 0.05)
                        No change in performance detected.
Found 82 outliers among 5000 measurements (1.64%)
  19 (0.38%) high mild
  63 (1.26%) high severe
```

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -28,10 +28,11 @@ debug = true
 ffi = ["lazy_static"]
 wasm_bindgen = ["js-sys", "wasm-bindgen"]
 python = ["pyo3"]
+full_bench = []
 
 [dependencies]
 abe_policy = "1.0"
-cosmian_crypto_core = "4.0"
+cosmian_crypto_core = { git = "https://github.com/Cosmian/crypto_core", rev = "0784e4e" }
```
Comment (Manuthor) [inline Cargo.toml:35]: just a comment to not forget to use the version instead the branch

**Review by Manuthor — COMMENTED** [review]
I know it is still a draft but it already looks good.

Do you think it could be useful to have in README.md a table representing the ciphertext once serialized? with 3 partitions and 10 partitions for example. A table more or less like this (this idea being to have a quick way to see the ciphertext size without going to `bib`):

| Number of partitions | ciphertext header size (in bytes) | ciphertext body size (in bytes) |
| --- | --- | --- |
| | | |

File: Cargo.toml [hunk 2]
```diff
@@ -28,10 +28,11 @@ debug = true
 ffi = ["lazy_static"]
 wasm_bindgen = ["js-sys", "wasm-bindgen"]
 python = ["pyo3"]
+full_bench = []
 
 [dependencies]
 abe_policy = "1.0"
-cosmian_crypto_core = "4.0"
```
Comment (ThibaudDauce) [inline Cargo.toml:34]: Need to release a new `crypto_core`?

**Review by ThibaudDauce — APPROVED** [review]
I didn't check the correctness of your length computations but it seems fine to me (did you try adding a `+1`/`-1`/`return 0` to the computed length to see if we have a failing test in this case?

**Comment by tbrezot** [comment]
> I didn't check the correctness of your length computations but it seems fine to me (did you try adding a `+1`/`-1`/`return 0` to the computed length to see if we have a failing test in this case?

I checked the lengths in the test `cover_crypt_core::tests::test_serialization()`. It fails when I add or remove 1 to the lengths.

**Review by Manuthor — APPROVED** [review]

**Review by Manuthor — APPROVED** [review]

**Pull request merged**